vertx-opentelemetry/src/main/java/io/vertx/tracing/opentelemetry/OpenTelemetryTracer.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (c) 2011-2021 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
+package io.vertx.tracing.opentelemetry;
+
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.Tracer;
+import io.opentelemetry.api.trace.propagation.W3CTraceContextPropagator;
+import io.opentelemetry.context.propagation.TextMapGetter;
+import io.opentelemetry.context.propagation.TextMapSetter;
+import io.vertx.core.Context;
+import io.vertx.core.spi.tracing.SpanKind;
+import io.vertx.core.spi.tracing.TagExtractor;
+import io.vertx.core.spi.tracing.VertxTracer;
+import io.vertx.core.tracing.TracingPolicy;
+
+import java.util.Map.Entry;
+import java.util.function.BiConsumer;
+
+import static io.opentelemetry.context.Context.current;
+import static io.vertx.tracing.opentelemetry.OpenTelemetryUtil.ACTIVE_CONTEXT;
+import static io.vertx.tracing.opentelemetry.OpenTelemetryUtil.ACTIVE_SPAN;
+
+public class OpenTelemetryTracer implements VertxTracer<Span, Span> {
+
+  private static final TextMapGetter<Iterable<Entry<String, String>>> getter = new HeadersPropagatorGetter();
+  private static final TextMapSetter<BiConsumer<String, String>> setter = new HeadersPropagatorSetter();
+
+  private final Tracer tracer;
+
+  OpenTelemetryTracer(final Tracer tracer) {
+    this.tracer = tracer;
+  }
+
+  @Override
+  public <R> Span receiveRequest(
+    final Context context,
+    final SpanKind kind,
+    final TracingPolicy policy,
+    final R request,
+    final String operation,
+    final Iterable<Entry<String, String>> headers,
+    final TagExtractor<R> tagExtractor) {
+
+    if (TracingPolicy.IGNORE.equals(policy)) {
+      return null;
+    }
+
+    final io.opentelemetry.context.Context parentContext = current();
+    final io.opentelemetry.context.Context tracingContext = W3CTraceContextPropagator.getInstance().extract(parentContext, headers, getter);
+
+    // OpenTelemetry SDK's Context is immutable, therefore if the extracted context is the same as the parent context
+    // there is no tracing data to propagate downstream and we can return null.
+    if (tracingContext == parentContext && TracingPolicy.PROPAGATE.equals(policy)) {
+      return null;
+    }
+
+    final Span span = tracer.spanBuilder(operation)
+      .setParent(tracingContext)
+      .setSpanKind(SpanKind.RPC.equals(kind) ? io.opentelemetry.api.trace.SpanKind.SERVER : io.opentelemetry.api.trace.SpanKind.CONSUMER)
+      .startSpan();
+
+    tagExtractor.extractTo(request, span::setAttribute);
```

Comment:
It sounds to me this change might require some modifications to the vertx tracing SPI? @vietj

Reply:
Currently the opentracing conventions are applied in other modules (e.g. Kafka module set its OT tags, see https://github.com/vert-x3/vertx-kafka-client/blob/master/src/main/java/io/vertx/kafka/client/common/tracing/TraceTags.java#L27-L31 ).
It's not ideal since it makes OT semantic conventions prevail on others, except for the "span.kind" tag which is set at this level. I don't know if the Otel conventions are the same or if they differ. If they don't differ, it should be fine, at least for a first PR.

But, perhaps later, I guess we should have a tags name mapping in Otel/OT/zipkin implementations.

Reply:
Thanks @jotak - yeah the conventions look mostly different. But I wouldn't worry on this first PR :)